File: backend/src/utils/pagination.ts
import type { Request } from "express";

export function getPagination(req: Request, defaultLimit = 20, maxLimit = 100) {
  const page = Math.max(1, Number(req.query.page) || 1);
  const limit = Math.min(maxLimit, Math.max(1, Number(req.query.limit) || defaultLimit));
  const skip = (page - 1) * limit;
  return { page, limit, skip, take: limit };
}

File: backend/app/core/pagination.py
from collections.abc import Callable
from dataclasses import dataclass

from fastapi import Query


@dataclass
class Pagination:
    page: int
    limit: int

    @property
    def skip(self) -> int:
        return (self.page - 1) * self.limit

    @property
    def take(self) -> int:
        return self.limit


def get_pagination(default_limit: int = 20, max_limit: int = 100) -> Callable[..., Pagination]:
    def dependency(
        page: int = Query(1, ge=1),
        limit: int = Query(default_limit, ge=1),
    ) -> Pagination:
        return Pagination(page=max(1, page), limit=min(max_limit, max(1, limit)))

    return dependency
